File: src/ui/composed.c
#include "kryon.h"
#include "ui_scaling.h"
#include "theme.h"


int CarouselControls(CarouselControlsProps p)
{
    if(p.count <= 0)
        return -1;

    int selected = ((p.selected % p.count) + p.count) % p.count;
    if(p.count == 1)
        return selected;
    if(!p.disabled)
        selected = (selected + p.move % p.count + p.count) % p.count;

    int hit = Scale(56);
    int inset = Scale(12);
    if(p.bounds.width >= hit * 2 && p.bounds.height >= hit) {
        for(int i = 0; i < 2; i++) {
            Rectangle bounds = {
                i ? p.bounds.x + p.bounds.width - inset - hit
                  : p.bounds.x + inset,
                p.bounds.y + (p.bounds.height - hit) / 2,
                hit,
                hit
            };
            int d = i ? 1 : -1;
            if(Button((ButtonProps){
                   .bounds = bounds,
                   .id = p.id + i,
                   .disabled = p.disabled,
                   .circle = true,
                   .icon_only = true,
                   .icon_type = i ? ICON_RIGHT : ICON_LEFT,
                   .tone = ButtonToneNeutral,
                   .emphasis = ButtonEmphasisSoft
               })) {
                selected = (selected + d + p.count) % p.count;
            }
        }
    }

    int indicator_hit = Scale(32);
    if(p.indicators.width >= indicator_hit * p.count &&
       p.indicators.height >= indicator_hit) {
        float x = p.indicators.x +
                  (p.indicators.width - indicator_hit * p.count) / 2;
        float y = p.indicators.y +
                  (p.indicators.height - indicator_hit) / 2;
        for(int i = 0; i < p.count; i++) {
            Rectangle bounds = {x + i * indicator_hit, y,
                                indicator_hit, indicator_hit};
            if(InvisibleButton((InvisibleButtonProps){
                   .bounds = bounds,
                   .id = p.id + 2 + i,
                   .disabled = p.disabled
               })) {
                selected = i;
            }
            if(IsWindowReady()) {
                DrawCircle((int)bounds.x + indicator_hit / 2,
                           (int)bounds.y + indicator_hit / 2,
                           Scale(i == selected ? 4 : 3),
                           Fade(GetThemeText(),
                                p.disabled ? 0.2f
                                           : (i == selected ? 1.0f : 0.28f)));
            }
        }
    }
    return selected;
}

File: go/kryon/composed.go
package kryon

type CarouselControlsProps struct {
	Bounds, Indicators    Rectangle
	Count, Selected, Move int32
	Disabled              bool
	ID                    int32
}

func CarouselControls(p CarouselControlsProps) int32 {
	if p.Count <= 0 {
		return -1
	}
	s := (p.Selected%p.Count + p.Count) % p.Count
	if p.Count == 1 {
		return s
	}
	if !p.Disabled {
		s = (s + p.Move%p.Count + p.Count) % p.Count
	}
	hit, inset := Scale(56), Scale(12)
	if p.Bounds.Width >= float32(hit*2) && p.Bounds.Height >= float32(hit) {
		for i := int32(0); i < 2; i++ {
			x := p.Bounds.X + float32(inset)
			if i == 1 {
				x = p.Bounds.X + p.Bounds.Width - float32(inset+hit)
			}
			b := Rectangle{X: x, Y: p.Bounds.Y + (p.Bounds.Height-float32(hit))/2, Width: float32(hit), Height: float32(hit)}
			d := int32(-1)
			iconType := int32(UIIconTypeLeft)
			if i == 1 {
				d = 1
				iconType = UIIconTypeRight
			}
			if Button(ButtonProps{
				Bounds: b, ID: p.ID + i, Disabled: p.Disabled,
				Circle: true, IconOnly: true, IconType: iconType,
				Tone: ButtonToneNeutral, Emphasis: ButtonEmphasisSoft,
			}) {
				s = (s + d + p.Count) % p.Count
			}
		}
	}
	hit = Scale(32)
	if p.Indicators.Width >= float32(hit*p.Count) && p.Indicators.Height >= float32(hit) {
		x := p.Indicators.X + (p.Indicators.Width-float32(hit*p.Count))/2
		y := p.Indicators.Y + (p.Indicators.Height-float32(hit))/2
		for i := int32(0); i < p.Count; i++ {
			b := Rectangle{X: x + float32(i*hit), Y: y, Width: float32(hit), Height: float32(hit)}
			if InvisibleButton(InvisibleButtonProps{Bounds: b, ID: p.ID + 2 + i, Disabled: p.Disabled}) {
				s = i
			}
			c := GetThemeText()
			r := int32(3)
			c.A = 71
			if i == s {
				r = 4
				c.A = 255
			}
			if p.Disabled {
				c.A = 51
			}
			Circle(int32(b.X+float32(hit)/2), int32(b.Y+float32(hit)/2), Scale(r), c)
		}
	}
	return s
}
